include/ginkgo/core/base/memory.hpp

```diff
@@ -0,0 +1,215 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_MEMORY_HPP_
+#define GKO_PUBLIC_CORE_BASE_MEMORY_HPP_
+
+
+#include <ginkgo/core/base/fwd_defs.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+namespace gko {
+
+
+/**
+ * Provides generic allocation and deallocation functionality to be used by an
+ * Executor.
+ */
+class Allocator {
+public:
+    virtual ~Allocator() = default;
+
+    virtual void* allocate(size_type num_bytes) const = 0;
+
+    virtual void deallocate(void* ptr) const = 0;
+};
+
+
+/**
+ * Implement this interface to provide an allocator for OmpExecutor or
+ * ReferenceExecutor.
+ */
+class CpuAllocatorBase : public Allocator {};
+
+
+/**
+ * Implement this interface to provide an allocator for CudaExecutor.
+ */
+class CudaAllocatorBase : public Allocator {};
+
+
+/**
+ * Implement this interface to provide an allocator for HipExecutor.
+ */
+class HipAllocatorBase : public Allocator {};
+
+
+/**
+ * Implement this interface to provide an allocator for DpcppExecutor.
+ */
+class DpcppAllocatorBase : public Allocator {
+public:
+    DpcppAllocatorBase(sycl::queue* queue);
+
+    void* allocate(size_type num_bytes) const final;
+
+    void deallocate(void* ptr) const final;
```

Comment:
only DPC++ has a queue that fully describes the device, for others, I need to set the environment from outside `allocate`, otherwise the interfaces would be more complex, and maybe do unnecessarily many cudaGetDevice calls. So I decided to only specialize the interface for DPC++.

Reply:
I removed the DPC++ allocator for now, since it requires more changes in the usage of queues.